### PR TITLE
Update README.md

### LINKED ISSUE #10: Write some basic user documentation for running simulations with plug-in
We should have some basic details in README of how to

 * Install plug-in (we can just point to FabSIM documentation for this)
 * Set required configuration options (for now just `neso_bin_dir`
 * Run a single job using `neso` task, and command line options available
 * Create a configuration directory for an ensembel run using the utility script
 * Run an ensemble job using `neso_ensemble` task, and command line options available

### PR DESCRIPTION
**Pull request by dleggat**
Installation and basic running examples for the FabNESO plugin.

Includes running a single simulation job with different solvers, using make_sweep_dir to create an ensemble directory, and the running of neso_ensemble for ensemble jobs.

Brief instructions are also given for running remotely, although we haven't finished setting that up entirely yet...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
 
 You will need [FabSim3 to be installed](https://fabsim3.readthedocs.io/en/latest/installation/) on the machine you will use to dispatch simulation runs from.
 
+NESO must be installed on the destination machine that will run the simulation.
```
Comment (matt-graham) [inline README.md]:
```suggestion
[NESO must be installed](https://github.com/ExCALIBUR-NEPTUNE/NESO#building-with-spack) on the destination machine that will run the simulation.
```

Possibly worth linking to NESO build instructions for completeness.

File: README.md [hunk 2]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
 ### Installation
 
+**Until the plugin is upstreamed to FabSIM properly, this first step is required:**
+
+Edit the file ` $FABSIM3_HOME/fabsim/deplot/plugins.yaml` to include the FabNESO plugin by including the following lines:
```
Comment (matt-graham) [inline README.md]:
```suggestion
Edit the file ` $FABSIM3_HOME/fabsim/deploy/plugins.yaml` to include the FabNESO plugin by including the following lines:
```

File: README.md [hunk 3]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+```
+FabNESO:
+  repository: https://github.com/UCL/fabneso.git
```
Comment (matt-graham) [inline README.md]:
```suggestion
  repository: https://github.com/UCL/FabNESO.git
```

Either capitalisation should work, but for consistency probably good to standardize on the `FabNESO` form.

File: README.md [hunk 4]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+```
+
+FabSIM can then automatically install the plugin by calling:
+
+`fabsim localhost install_plugin:FabNESO`
```
Comment (matt-graham) [inline README.md]:
````suggestion
```
fabsim localhost install_plugin:FabNESO
```
````

File: README.md [hunk 5]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+Before the code can be run, a file `machines_FabNESO_user.yml` must be created in the plugin's directory containing the paths to NESO installations on each system that will be used for running.
+An example file `machines_FabNESO_user_example.yml` is provided to use as a template.
+
 ### Running simulations locally
 
+NESO runs by calling the desired solver with a `conditions` and `mesh` file that define the parameters of the simulation.
```
Comment (matt-graham) [inline README.md]:
```suggestion
NESO runs by calling the desired solver with a conditions and mesh XML files that define the parameters of the simulation and geometry of the simulation domain respectively.
```

File: README.md [hunk 6]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+Examples of these configuration files are provided in the `config_files/two_stream/` (intended for use with NESO's Electrostatic2D3V solver) and `config_files/2Din3D-hw/` (for use with the H3LAPD solver) directories.
```
Comment (matt-graham) [inline README.md]:
```suggestion
Examples of these configuration files are provided in the directories `config_files/two_stream/` (intended for use with NESO's `Electrostatic2D3V` solver) and `config_files/2Din3D-hw/` (for use with the `H3LAPD` solver).
```

Moving `directories` forwards makes it a bit clearer to me both subsequent clauses are included in this to me.

File: README.md [hunk 7]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+#### Running a single simulation
+
+The FabSIM [`neso`](https://github.com/UCL/fabneso/blob/89bba5ca73c333cf16e8b8dddee164a67b1eb867/FabNESO.py#L20) task runs a single simulation on the machine of your chosing.
```
Comment (matt-graham) [inline README.md]:
```suggestion
The FabSIM `neso` task runs a single simulation on the machine of your chosing.
```
Not a bad idea to link to the function definition and docstring, but using a specific commit hash will mean this quickly will refer to an out of date version (and using the latest commit on `main` will mean the line reference probably ends up pointing to the wrong place). We can build HTML documentation from the docstrings and host on GitHub Pages and then link there in future (will open a separate issue for this).

File: README.md [hunk 8]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+To run NESO locally using the two_stream example, run the following command:
```
Comment (matt-graham) [inline README.md]:
```suggestion
To run NESO locally using the `Electrostatic2D3V` solver with [the `two_stream` example](https://github.com/ExCALIBUR-NEPTUNE/NESO/tree/main/examples/Electrostatic2D3V/two_stream), run the following command:
```

File: README.md [hunk 9]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+```
+fabsim localhost neso:two_steam
```
Comment (matt-graham) [inline README.md]:
```suggestion
fabsim localhost neso:two_stream
```

File: README.md [hunk 10]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+```
+
+The first positional argument after `neso:` specifies the directory within `config_files` that holds the conditions and mesh files to be used.
+
+Additional arguments that can be given the to the `neso` task are:
+
+- `solver` : chose which NESO solver to run. Default = `Electrostatic2D3V`,
+- `conditions_file_name` : Name of conditions XML file in configuration directory. Default = `conditions.xml`,
+- `mesh_file_name` : Name of mesh XML in configuration directory. Default = `mesh.xml`
```
Comment (matt-graham) [inline README.md]:
```suggestion
- `mesh_file_name` : Name of mesh XML file in configuration directory. Default = `mesh.xml`.
```

File: README.md [hunk 11]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+```
+fabsim localhost neso:two_steam
+```
+
+The first positional argument after `neso:` specifies the directory within `config_files` that holds the conditions and mesh files to be used.
+
+Additional arguments that can be given the to the `neso` task are:
+
+- `solver` : chose which NESO solver to run. Default = `Electrostatic2D3V`,
+- `conditions_file_name` : Name of conditions XML file in configuration directory. Default = `conditions.xml`,
```
Comment (matt-graham) [inline README.md]:
```suggestion
- `conditions_file_name` : Name of conditions XML file in configuration directory. Default = `conditions.xml`.
```

File: README.md [hunk 12]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+```
+fabsim localhost neso:two_steam
+```
+
+The first positional argument after `neso:` specifies the directory within `config_files` that holds the conditions and mesh files to be used.
+
+Additional arguments that can be given the to the `neso` task are:
+
+- `solver` : chose which NESO solver to run. Default = `Electrostatic2D3V`,
```
Comment (matt-graham) [inline README.md]:
```suggestion
- `solver` : chose which NESO solver to run. Default = `Electrostatic2D3V`.
```

File: README.md [hunk 13]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+- `conditions_file_name` : Name of conditions XML file in configuration directory. Default = `conditions.xml`,
+- `mesh_file_name` : Name of mesh XML in configuration directory. Default = `mesh.xml`
+
+To run the H3LAPD solver with the example configurations, for example, the following command should be run:
```
Comment (matt-graham) [inline README.md]:
```suggestion
To run the H3LAPD solver with [the `2Din3D-hw` example configuration](https://github.com/ExCALIBUR-NEPTUNE/NESO/tree/main/examples/H3LAPD/2Din3D-hw), for example, the following command should be run:
```

File: README.md [hunk 14]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+To retrieve the results of a job, run the command:
+
+```
+fabsim localhost fetch_results
+```
+
+#### Running an ensemble of simulations
+
+The FabSIM [`neso_ensemble`](https://github.com/UCL/fabneso/blob/89bba5ca73c333cf16e8b8dddee164a67b1eb867/FabNESO.py#L49) task runs a series of FabSIM jobs taking a `SWEEP' directory as its primary input.
```
Comment (matt-graham) [inline README.md]:
```suggestion
The FabSIM `neso_ensemble` task runs a series of FabSIM jobs taking a `SWEEP' directory as its primary input.
```

File: README.md [hunk 15]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+This SWEEP directory contains any number of subdirectories, each containing a conditions and mesh file for individual NESO jobs.
```
Comment (matt-graham) [inline README.md]:
```suggestion
This `SWEEP` directory contains any number of subdirectories, each containing a conditions and mesh file for individual NESO jobs.
```

File: README.md [hunk 16]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+A utility script [utils/make_sweep_dir.py](https://github.com/UCL/FabNESO/blob/main/utils/make_sweep_dir.py) is provided to automatically build this sweep directory and encode the input conditions files with templated parameters selected by the user.
```
Comment (matt-graham) [inline README.md]:
```suggestion
A utility script [`utils/make_sweep_dir.py`](https://github.com/UCL/FabNESO/blob/main/utils/make_sweep_dir.py) is provided to automatically build this sweep directory and encode the input conditions files with templated parameters selected by the user.
```

File: README.md [hunk 17]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+An example use to template the `particle_initial_velocity` parameter for 4 values between 0.1 and 2.0 using the the two_stream example files would therefore be:
+
+```
+python3 make_sweep_dir.py --para_to_template="particle_initial_velocity" --scan_min=0.1 --scan_max=2.0 --n_divs=4
+```
+
+The script can also template an arbitrary number of parameters in the conditions file using the following command line argument:
+
+- `--parameter_dict` : A python dict of the parameters to be scanned, with associated minimum and maximum values as a list (default = `""`)
```
Comment (matt-graham) [inline README.md]:
```suggestion
- `--parameter_dict` : A Python dict of the parameters to be scanned, with associated minimum and maximum values as a list (default = `""`)
```

File: README.md [hunk 18]
```diff
@@ -32,12 +32,136 @@ Centre for Advanced Research Computing, University College London ([arc.collabor
...
+
+```
+fabsim localhost neso_ensemble:two_stream_ensemble
+```
+
+`neso_ensemble` takes the additional input parameters:
+
+- `solver` : chose which NESO solver to run. Default = `Electrostatic2D3V`,
+- `conditions_file_name` : Name of conditions XML file in SWEEP directories. Default = `conditions.xml`,
+- `mesh_file_name` : Name of mesh XML in the SWEEP directories. Default = `mesh.xml`
```
Comment (matt-graham) [inline README.md]:
```suggestion
- `mesh_file_name` : Name of mesh XML file in the SWEEP directories. Default = `mesh.xml`
```

**Review by matt-graham — CHANGES_REQUESTED** [review]
Thanks - this looks great in general. Have made some suggested fixes for a few typos and some minor formatting changes.

File: README.md (hunk 7 above)
Comment (dleggat) [reply README.md]: Ah okay, yes that sounds like a much more sensible way of linking it.

**Review by matt-graham — APPROVED** [review]
Thanks for the updates, all looks good to merge to me!

**Pull request merged**